Deprecate ACTIVE_HUNT_ID env variable

akirabaruah opened the pull request:
Rather than relying on an app environment variable, we can use the
boolean "active" field on Hunt objects in the database to find the
current active hunt.

Removes all uses of ACTIVE_HUNT_ID and updates dependent logic to use a
database query for active=True instead.

This will help in supporting true multi-hunt functionality (see https://github.com/cardinalitypuzzles/smallboard/issues/156).

akirabaruah commented:
@erwa I think you originally added ACTIVE_HUNT_ID, so I added you as a reviewer (other folks, feel free to review as well :smile:)

rgossiaux commented on hunts/views.py:
```diff
@@ -214,5 +216,20 @@ def post(self, request, pk):
+
+    def get_redirect_url(self, *args, **kwargs):
+        try:
+            active_hunt = Hunt.objects.get(active=True)
```
I think this is going to raise an exception when there are multiple hunts marked active (my local db is in this state for example).

What do you think about keeping track of it per-user? Maybe it's overkill but we could maybe then update it on each page load, so whatever hunt you were last looking at is your default.

akirabaruah replied:
Ooh yeah that sounds good. So, if the user's never opened smallboard we can load a generic home page (probably needed for https://github.com/cardinalitypuzzles/smallboard/issues/197 anyway as @rawxfish pointed out), otherwise redirect to the last accessed hunt. To achieve this, I'm thinking we'll want to:

- add an `active_hunt = ForeignKey(Hunt, null=True)` to the `Puzzler` user model
- update `Puzzler.active_hunt` whenever a user accesses a `Hunt`
- redirect based on `Puzzler.active_hunt` or load the home page if it's null
- remove the `Hunt.active` field

Am I missing anything?

erwa replied:
Some related thoughts: It'd be nice if there's somewhere users can toggle between hunts if they are working on multiple hunts simultaneously.

Also, we might consider having unique URLs per hunt so you can differentiate hunts just by looking at that. E.g.: hunt/mitmh2021/... vs. hunt/crums2020/... etc.

Another use case to think about supporting in the future (which has been discussed elsewhere) is having multiple instances of the same hunt, so different teams can use smallboard for the same hunt and not affect each other or see each other's answers. Maybe this could be accomplished by introducing the notion of a Team and having puzzlers and answer submissions associated with teams. Just something to think about.

akirabaruah replied:
> Some related thoughts: It'd be nice if there's somewhere users can toggle between hunts if they are working on multiple hunts simultaneously.

Maybe something like a user settings page? Filed https://github.com/cardinalitypuzzles/smallboard/issues/203

> Also, we might consider having unique URLs per hunt so you can differentiate hunts just by looking at that. E.g.: hunt/mitmh2021/... vs. hunt/crums2020/... etc.

Do you mean a site URL per hunt? In other words, maybe a separate page per hunt?

Or do you mean enforcing a unique *external* URL per hunt?

> Another use case to think about supporting in the future (which has been discussed elsewhere) is having multiple instances of the same hunt, so different teams can use smallboard for the same hunt and not affect each other or see each other's answers. Maybe this could be accomplished by introducing the notion of a Team and having puzzlers and answer submissions associated with teams. Just something to think about.

:+1: My thoughts exactly (and @rgossiaux's too, I think). See https://github.com/cardinalitypuzzles/smallboard/issues/191 for the "elsewhere" discussions :smile:

Are any changes we should make in this PR to address the different instances per Team?

rgossiaux replied:
@akirabaruah those changes sound good to me, though you could probably leave Hunt.active as it is if you wanted; not really sure about the context but I'm guessing the idea was to distinguish between hunts in-progress and completed hunts which is a little different from "which hunt a user is/was working on".

Really like the unique URL per hunt, that's a lot more readable than /2/ and stuff lol

rgossiaux commented:
thanks for getting rid of this, it helps a lot with setup/config imo :)

erwa commented:
> @erwa I think you originally added ACTIVE_HUNT_ID, so I added you as a reviewer (other folks, feel free to review as well 😄)

Actually, @rawxfish originally added it in 914128a5 :).

erwa commented on smallboard/urls.py:
```diff
@@ -19,15 +19,16 @@
+    path("accounts/", include("django.contrib.auth.urls")),
+    path("accounts/", include("accounts.urls")),
+    path("puzzles/", include("puzzles.urls")),
+    path("hunts/", include("hunts.urls", namespace="hunts")),
+    path("answers/", include("answers.urls")),
```
Is the namespace="hunts" part necessary? Or else, should we also add namespace="answers" for consistency?

erwa reviewed (commented):
This looks like a step in the right direction. Have a few comments and questions.

akirabaruah commented:
Added a slug field for hunts so that they're human readable. I ended up leaving the Hunt.active field as @rgossiaux suggested and added a last_accessed_hunt to the User model to clarify the field's intended purpose.

akirabaruah reviewed (commented):
Added a slug field for hunts so that they're human readable. I ended up leaving the Hunt.active field as @rgossiaux suggested and added a last_accessed_hunt to the User model to clarify the field's intended purpose.

erwa commented on answers/urls.py:
```diff
@@ -3,10 +3,13 @@
-    path("queue/<int:hunt_pk>", AnswerView.as_view()),
-    path("queue/<int:hunt_pk>/answers", answers),
-    path("queue/<int:hunt_pk>/<int:answer_pk>", AnswerView.as_view()),
+    path("queue/<slug:hunt_slug>", AnswerView.as_view(), name="hunt_queue"),
```
How about `submission_queue` for the name?

erwa reviewed (approved):
I like the namespacing of URLs -- makes it more clear what the URLs are for. Also, thanks for adding the hunt slug field!

akirabaruah commented on hunts/views.py:
```diff
@@ -214,5 +216,23 @@ def post(self, request, pk):
+        if not hunt:
+            return reverse('hunts:index')
+
+        # Answer urls use pattern_name="answers:hunt_queue" and "hunt_slug" to
+        # disambiguate between answer and hunt primary keys.
+        if self.pattern_name == LastAccessedHuntRedirectView.pattern_name:
+            kwargs['slug'] = hunt.slug
+        else:
+            kwargs['hunt_slug'] = hunt.slug
```
As far as I can tell, within the `hunt` app, it was referred to as "id" whereas in the `answers` app, it was called "hunt_id" to disambiguate from Answer "id" values. "hunt_id" / "hunt_slug" seems *slightly* redundant within the `hunt` app, but it seems worth using everywhere to reduce this RedirectView's complexity :+1:

akirabaruah commented:
I think I've incorporated everyone's feedback so far (barring anything we deferred to newly filed Issues. Please let me know if I've missed anything - thanks!

Tests to follow :smile:

akirabaruah commented:
> Tests to follow smile

Actually, I'm gonna merge this as is, then follow up with tests after getting some coverage metrics. Otherwise, it's a bit unclear when we've added a sufficient level of testing.
